web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php

```diff
@@ -0,0 +1,357 @@
+    // early. The entity is already processed.
+    elseif ($this->hasSolutionDataChildDependency($parent, $entity)) {
+      return;
+    }
+
+    // If the entity is a solution, do not descend further because each solution
+    // has its own dependencies. Related solutions do not affect the solution
+    // itself.
+    elseif ($entity->bundle() === 'solution') {
+      return;
+    }
+
+    // This bundle has no entity reference fields.
+    if (!$reference_fields = array_keys($this->getAdmsSchemaEntityReferenceFields($entity->bundle(), ['rdf_entity']))) {
+      return;
+    }
+
+    foreach ($reference_fields as $field_name) {
+      /** @var \Drupal\Core\Field\EntityReferenceFieldItemListInterface $field */
+      $field = $entity->get($field_name);
+
+      foreach ($this->getStagingReferencedEntities($field) as $referenced_entity) {
+        if ($this->hasSolutionDataChildDependency($parent, $referenced_entity)) {
+          // The entity has already been processed.
+          continue;
+        }
+
+        // Only for the case of the release, dive deeper and check whether there
+        // are more referenced entities. Normally, the only difference between
+        // the solution and the release, would be the distributions.
+        if ($referenced_entity->bundle() === 'asset_release') {
+          $this->buildSolutionDependencyTree($referenced_entity, $parent);
+        }
+
+        // Do not add solutions and licences as a dependency. Solutions are a
+        // tree on their own and licences are stored in Joinup and are not
+        // imported.
+        if (!in_array($referenced_entity->bundle(), ['licence', 'solution'])) {
+          $this->addSolutionDataChildDependency($parent, $referenced_entity);
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns a list of entity IDs, given a reference field.
+   *
+   * @param \Drupal\Core\Field\EntityReferenceFieldItemListInterface $field
+   *   The entity reference field.
+   *
+   * @return \Drupal\Core\Entity\EntityInterface[]
+   *   An array of entities referenced.
+   */
+  protected function getStagingReferencedEntities(EntityReferenceFieldItemListInterface $field): array {
+    $ids = [];
+    if (!$field->isEmpty()) {
+      // Can't use EntityReferenceFieldItemListInterface::referencedEntities()
+      // here because that doesn't filter on 'staging' graph.
+      $ids = array_filter(array_unique(array_map(function (array $item): string {
+        return $item['target_id'];
+      }, $field->getValue())));
+    }
+    return Rdf::loadMultiple($ids, ['staging']);
+  }
+
+  /**
+   * Builds an array of categories for solutions stored in the dependency tree.
+   *
+   * @param array $solution_ids
+   *   A list of solution ids for which to calculate the category.
+   */
+  protected function buildSolutionsCategories(array $solution_ids): void {
+    foreach ($solution_ids as $parent) {
+      $provenance_record = $this->provenanceHelper->loadOrCreateEntityActivity($parent);
```

Comment:
I was thinking that if a solution changes, then a change will most probably apply to the solution itself. For example, even if there is no change to the description, if a new release of the solution is created, then the `hasVersion` property of the solution will retrieve this release. Meaning that the solution hash will change anyway. So in real case scenarios, it will be rare that a solution will not change but a sub entity will. These would be cases like a contact information has changed or a distribution description is changed.
In all other cases, changes will probably propagate to the parent solution.
In most common cases, we avoid loading activities for all sub entities.

Reply:
On second thought, this is not relevant. I do not actually check whether the solution is already changed here. I thought I did. However, I am only passing entity ids around. I can load them and convert the `::isSolutionChanged` to something like `::containsChangedRecords` and pass a list of provenance records to compare against the current solution data.